File: leetcode/lc/T301-350/t_312.java
package leecoder;

import leetcoder.ListNode;

import java.util.*;
import java.util.stream.Collectors;

class Solution {
    public int maxCoins(int[] nums) {
        int n = nums.length;
        int []no = new int[n+2];
        for (int i = 0; i < n; i++) {
            no[i+1] = nums[i];
        }
        no[0] = no[n+1] = 1;
        int [][] dp = new int[n+2][n+2];
        // k 1-n
        // i+k 1-n+1
        // j = i+ k -1  // j 0-n
        //dp[i][j] = dp[i][x-1] + dp[x+1][j] + no***;

        for (int k = 1; k <=n ; k++) {
            for (int i = 1; i <=n-k+1 ; i++) {
                int j = i+k-1;
                for (int x = i; x <=j ; x++) {
                    dp[i][j] = Math.max(dp[i][j],
                            dp[i][x-1]+dp[x+1][j] + no[i-1]*no[x]*no[j+1]);
                }

            }
        }
        return dp[1][n];

    }
}


public class Main {
    public static void main(String[] args) {
        int i = new leecoder.Solution().maxCoins(new int[]{3, 1, 5, 8});
        System.out.println(i);
    }
}

File: leetcode/lc/T301-350/T312.cc
#include <bits/stdc++.h>
using namespace std;

class Solution
{
public:
    int maxCoins(vector<int> nums)
    {
        vector<int> v;
        v.emplace_back(1);
        for (auto it : nums)
            v.emplace_back(it);
        v.emplace_back(1);

        int len = v.size();
        vector<vector<int>> dp(len, vector<int>(len, 0));
        for (int i = nums.size(); i >= 0; i--)
            for (int j = i + 1; j < len; j++)
                for (int k = i + 1; k < j; k++)
                {
                    dp[i][j] = max(dp[i][j], dp[i][k] + dp[k][j] + v[i] * v[k] * v[j]);
                }

        return dp[0][len - 1];
    }
};
int main()
{
    Solution s;

    cout
        << s.maxCoins({3, 1, 5, 8}) << endl;

    int n = 5;
    // 斜着遍历数组
    for (int l = 2; l <= n; l++)
    {
        for (int i = 0; i <= n - l; i++)
        {
            int j = l + i - 1;
            // 计算 dp[i][j]
            cout << i << " " << j << " ";
        }
        cout << endl;
    }
}